kernel/libs/aster-bigtcp/src/socket/bound.rs

```diff
@@ -1,94 +1,124 @@
     sync::atomic::{AtomicBool, AtomicU64, AtomicU8, Ordering},
 };
 
-use ostd::sync::{LocalIrqDisabled, RwLock, SpinLock, SpinLockGuard, WriteIrqDisabled};
+use keyable_arc::KeyableArc;
+use ostd::sync::{LocalIrqDisabled, SpinLock, SpinLockGuard};
 use smoltcp::{
     iface::Context,
     socket::{tcp::State, udp::UdpMetadata, PollAt},
     time::{Duration, Instant},
-    wire::{IpAddress, IpEndpoint, IpRepr, TcpControl, TcpRepr, UdpRepr},
+    wire::{IpEndpoint, IpRepr, TcpControl, TcpRepr, UdpRepr},
 };
+use spin::Once;
+use takeable::Takeable;
 
 use super::{
     event::{SocketEventObserver, SocketEvents},
-    option::RawTcpSetOption,
+    option::{RawTcpOption, RawTcpSetOption},
+    unbound::{new_tcp_socket, new_udp_socket},
     RawTcpSocket, RawUdpSocket, TcpStateCheck,
 };
-use crate::{ext::Ext, iface::Iface};
+use crate::{
+    ext::Ext,
+    iface::{BindPortConfig, BoundPort, Iface},
+};
 
-pub struct BoundSocket<T: AnySocket<E>, E: Ext>(Arc<BoundSocketInner<T, E>>);
+pub struct Socket<T: Inner<E>, E: Ext>(Takeable<KeyableArc<SocketBg<T, E>>>);
 
-/// [`TcpSocket`] or [`UdpSocket`].
-pub trait AnySocket<E> {
-    type RawSocket;
-    type Observer: SocketEventObserver;
+impl<T: Inner<E>, E: Ext> PartialEq for Socket<T, E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.eq(&other.0)
+    }
+}
+impl<T: Inner<E>, E: Ext> Eq for Socket<T, E> {}
+impl<T: Inner<E>, E: Ext> PartialOrd for Socket<T, E> {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl<T: Inner<E>, E: Ext> Ord for Socket<T, E> {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+impl<T: Inner<E>, E: Ext> Borrow<KeyableArc<SocketBg<T, E>>> for Socket<T, E> {
+    fn borrow(&self) -> &KeyableArc<SocketBg<T, E>> {
+        self.0.as_ref()
+    }
+}
 
-    /// Called by [`BoundSocket::new`].
-    fn new(socket: Box<Self::RawSocket>) -> Self;
+/// [`TcpConnectionInner`] or [`UdpSocketInner`].
+pub trait Inner<E: Ext> {
+    type Observer: SocketEventObserver;
 
-    /// Called by [`BoundSocket::drop`].
-    fn on_drop(this: &Arc<BoundSocketInner<Self, E>>)
+    /// Called by [`Socket::drop`].
+    fn on_drop(this: &KeyableArc<SocketBg<Self, E>>)
     where
         E: Ext,
         Self: Sized;
 }
 
-pub type BoundTcpSocket<E> = BoundSocket<TcpSocket, E>;
-pub type BoundUdpSocket<E> = BoundSocket<UdpSocket, E>;
+pub type TcpConnection<E> = Socket<TcpConnectionInner<E>, E>;
+pub type TcpListener<E> = Socket<TcpListenerInner<E>, E>;
+pub type UdpSocket<E> = Socket<UdpSocketInner, E>;
 
-/// Common states shared by [`BoundTcpSocketInner`] and [`BoundUdpSocketInner`].
-pub struct BoundSocketInner<T: AnySocket<E>, E> {
-    iface: Arc<dyn Iface<E>>,
-    port: u16,
-    socket: T,
-    observer: RwLock<T::Observer, WriteIrqDisabled>,
+/// Common states shared by [`TcpConnectionBg`] and [`UdpSocketBg`].
+pub struct SocketBg<T: Inner<E>, E: Ext> {
```

Comment:
What does `Bg` mean in this type?

Reply:
This means "background". A foreground socket (e.g. `TcpConnection`) handles system calls from the user program and a background socket (e.g. `TcpConnectionBg`) handles packets from the network.

Do you have better naming suggestions?

Reply:
How about naming it SocketWorker? This suggestion comes from GPT. The term `Background` gives me the impression of a background process.

Another option could be `SocketBackend`, but I'm unsure if that might lead to confusion. In virtualization scenarios, we often refer to virtio-net as the frontend and TAP as the backend.

Reply:
Well, my first initiative to "worker" is a worker process. So I don't think it's very ideal either.

Reply:
Since we can't think up any better name, `Bg` is also acceptable for me and we can keep the current name.

Reply:
I also added some doc comments to explain what `Bg` means in the type name.